README.md
# NRR (node redis RPC)

Simple rpc for node using redis. This library extends [node redis pubsub](https://www.npmjs.com/package/node-redis-pubsub)
and uses its feature as a base. It uses the existing methods `on/subscribe` / `emit/publish` to realize a simple rpc
over redis and between node.js instances. Every node.js instance can connect to the event bus, listen to the bus via
pub/sub or register/trigger rpc calls.


[![npm package](https://img.shields.io/npm/v/node-redis-rpc.svg?style=flat-square)](https://www.npmjs.org/package/node-redis-rpc)
[![Dependency Status](https://david-dm.org/Bruce17/node-redis-rpc.svg?style=flat-square)](https://david-dm.org/Bruce17/node-redis-rpc)
[![devDependency Status](https://david-dm.org/Bruce17/node-redis-rpc/dev-status.svg?style=flat-square)](https://david-dm.org/Bruce17/node-redis-rpc#info=devDependencies)
[![code climate](https://img.shields.io/codeclimate/github/Bruce17/node-redis-rpc.svg?style=flat-square)](https://codeclimate.com/github/Bruce17/node-redis-rpc)
[![Coverage Status](https://coveralls.io/repos/github/Bruce17/node-redis-rpc/badge.svg?branch=master)](https://coveralls.io/github/Bruce17/node-redis-rpc?branch=master)
[![Travis CI](https://travis-ci.org/Bruce17/node-redis-rpc.svg?style=flat-square)](https://travis-ci.org/Bruce17/node-redis-rpc)


## Important Change

See [Changelog](Changelog.md) for more information.


## Install

```bash
$ npm install node-redis-rpc
```

## Usage
### Setup

```javascript
var NodeRedisRpc = require('node-redis-rpc');
var config = {
    host: 'localhost', // redis server hostname
    port: 6379,        // redis server port
    auth: 'password',  // optional password
    scope: 'test'      // use scope to prevent sharing messages between "node redis rpc"
};
var nodeRedisRpcInst = new NodeRedisRpc(config);
```

### Simple rpc

```javascript
// Register a listener on the channel "foo:bar" [1]
nodeRedisRpcInst.on('foo:bar', function (data, channel, done) {
    // do s.th. ...
    
    // Trigger done handler to fire back rpc result [2]
    // - first arg:  error status
    // - second arg: result data
    done(null, {num: 123, ary: [1,2,3,4], text: 'hello'});
});

/**
 * RPC callback handler [3]
 *
 * @param {null|*} err    error status
 * @param {*}      result result data returned by the rpc callback (see [2])
 */
var myRpcCallback = function (err, result) {
    console.log('err', err);       // outputs: 'null'
    console.log('result', result); // outputs: '{foo: 'bar', num: 123}'
};

// Trigger an event on the channel "foo:bar" (received by [1])
nodeRedisRpcInst.emit(
    'foo:bar',      // channel
    {name: 'Hans'}, // message data
    {               // options
        type: 'rpc',            // trigger an event of type "rpc"
        callback: myRpcCallback // register a callback handler [3] to be executed when the rpc result returns
    }
);
```

### More

Please see [node redis pubsub](https://www.npmjs.com/package/node-redis-pubsub) for more information on how to use
this package's base package e.g. unsubscrbe to a channel.


CHANGELOG.md
Changelog
=========

## 2.0.0
* update all dependencies to latest version
* removed old Node.js versions, minimum requirement is now Node.js v4.2

## 1.0.8
* update `README.md` & `.npmignore` file

## 1.0.7
* updated dependencies

## 1.0.6
* removed automatic npm publish from travis

## 1.0.5
* updated gruntfile build process, added more comments

## 1.0.4
* added redis server to the travis build environment (for tests only)

## 1.0.3
* adjusted travis config and unit tests

## 1.0.2
* minor changes, added status icons, added travis support

## 1.0.0
* first commit


test/node-redis-rpc.spec.js
var chai = require('chai');
var sinon = require('sinon');
var sinonChai = require('sinon-chai');
var expect = chai.expect;
//chai.should();
chai.use(sinonChai);

var NodeRedisRpc = require('../index');

var conf = {port: 6379, scope: 'onescope'};
var conf2 = {port: 6379, scope: 'anotherscope'};

describe('Node Redis RPC', function () {
    var stubs = {};
    var nrr

    beforeEach(function () {
        stubs.consoleInfo = sinon.stub(console, 'info');
        
        nrr = null
    });

    afterEach(function () {
        var key;
        for (key in stubs) {
            if (stubs.hasOwnProperty(key)) {
                stubs[key].restore();
            }
        }
        
        // Safely (connections will be closed properly once all commands are sent)
        nrr.quit();
    });
    
    after(function () {
        // Dangerously (connections will be immediately terminated)
        nrr.end();
    })

    // NOTICE: copied from "node redis pubsub"
    it('Should send and receive standard messages correctly', function (done) {
        nrr = new NodeRedisRpc(conf);

        nrr.on('a test #1',
            function (data, channel) {
                data.first.should.equal('First message');
                data.second.should.equal('Second message');
                channel.should.equal('onescope:a test #1');
                done();
            },
            function () {
                nrr.emit('a test #1', {
                    first: 'First message',
                    second: 'Second message'
                });
            }
        );
    });

    it('Should send and receive rpc messages correctly', function (done) {
        nrr = new NodeRedisRpc(conf);

        nrr.on('a test #2',
            function (data, channel, rpcDone) {
                data.first.should.equal('First message');
                data.second.should.equal('Second message');
                channel.should.equal('onescope:a test #2');

                rpcDone(null, {rpcData: 'test test'});
            },
            function () {
                nrr.emit('a test #2',
                    {
                        first: 'First message',
                        second: 'Second message'
                    },
                    {
                        type: 'rpc',
                        callback: function (err, result) {
                            expect(err).to.be.null;
                            expect(result).to.be.ok;
                            result.rpcData.should.equal('test test');

                            done();
                        }
                    }
                );
            }
        );
    });
});


Gruntfile.js
/**
 * @author Michael Raith
 * @email  michael.raith@bcmsolutions.de
 * @date   21.04.2015 10:37
 */

/*eslint-disable */

module.exports = function (grunt) {
    'use strict';

    require('load-grunt-tasks')(grunt); // npm install --save-dev load-grunt-tasks

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),

        /**
         * Some basic information e.g. to create a banner
         */
        basicInformation: {
            author: 'Michael Raith <mia87@web.de>',
            url: 'https://github.com/Bruce17',
            date: '<%= grunt.template.today("yyyy-mm-dd HH:MM:ss") %>'
        },

        banner: '/**\n' +
        ' * @author <%= basicInformation.author %>\n' +
        ' * @url    <%= basicInformation.url %>\n' +
        ' * @date   <%= basicInformation.date %>\n' +
        ' */\n',

        files: {
            js: [
                {
                    expand: true,
                    cwd: 'src/',
                    src: '**/*.js',
                    dest: 'lib/',
                    filter: 'isFile'
                }
            ],
            jsLib: [
                {
                    expand: true,
                    cwd: 'lib/',
                    src: '**/*.js',
                    dest: 'lib/',
                    filter: 'isFile'
                }
            ],
            test: [
                'test/**/*.spec.js'
            ]
        },


        /**
         * Do some code style checks.
         */
        eslint: {
            options: {
                configFile: '.eslintrc'
            },
            target: [
                'src/**/*.js'
            ]
        },


        /**
         * Copy files.
         */
        copy: {
            main: {
                files: '<%= files.js %>'
            }
        },


        /**
         * Remove console.logs statements.
         */
        removelogging: {
            main: {
                options: {
                    namespace: [
                        'console'
                    ],
                    methods: [
                        'log',
                        'info',
                        'warn',
                        'debug',
                        'error'
                    ],
                    replaceWith: ''
                },

                files: '<%= files.jsLib %>'
            }
        },


        /**
         * Add a banner to all created lib files.
         */
        usebanner: {
            main: {
                options: {
                    position: 'top',
                    banner: '<%= banner %>',
                    linebreak: true
                },

                files: '<%= files.jsLib %>'
            }
        },


        /**
         * Test files
         */
        mochaTest: {
            test: {
                options: {
                    reporter: 'spec',
                    require: [
                        'should'
                    ],
                    timeout: 20000 // 20 seconds
                },
                src: '<%= files.test %>'
            }
        }
    });

    grunt.registerTask('test', ['mochaTest:test']);

    // Default tasks
    grunt.registerTask('default', ['eslint', 'test', 'copy:main', 'removelogging:main', 'usebanner:main']);
};

src/node-redis-rpc.js
var NRP = require('node-redis-pubsub');
var nodeUuid = require('uuid');


/**
 * Create a new NodeRedisRpc instance that can subscribe to channels and publish messages.
 *
 * @inheritDoc
 * @constructor
 */
function NodeRedisRpc () {
    'use strict';

    //// Inherit from "node redis pubsub" and extend everything needed for RPC ...
    NRP.apply(this, arguments);
}
NodeRedisRpc.prototype = NRP.prototype;


// Extend the original "publish" and call it later ...
NodeRedisRpc.prototype._emit = NodeRedisRpc.prototype.emit;

/**
 * Emit an (rpc) event
 *
 * @param {String} channel Channel on which to emit the message
 * @param {Object} message
 * @param {Object} options Optional options
 */
NodeRedisRpc.prototype.emit = function (channel, message, options) {
    'use strict';

    options = options || {};
    var rpcCallback = options.callback || function () {};

    // Extend original "publish" method to work for rpc calls
    if (options.type === 'rpc') {
        var self = this;
        var uuid = nodeUuid.v4();

        // Tell the responding method that we're executing a rpc request and want to receive a response.
        message.__type = 'rpc';
        message.__backChannel = uuid;

        console.info(
            'node redis rpc\n' +
            '\t>> emit\n' +
            '\t\ttype: %s\n' +
            '\t\tuuid: %s\n' +
            '\t\tcallback: %s',
            message.__type,
            message.__backChannel,
            rpcCallback
        );

        // Register a callback handler for the rpc response.
        // Remove handler after response was received.
        self.on(uuid, function (data) {
            rpcCallback(data.err, data.result);

            // Unsubscribe rpc response handler
            self.off(uuid);
        });
    }

    // Trigger original method
    this._emit(channel, message);
};


// Extend the original "subscribe" and call it later ...
NodeRedisRpc.prototype._on = NodeRedisRpc.prototype.on;

/**
 * @inheritDoc
 */
NodeRedisRpc.prototype.on = function (channel, handler, callback) {
    'use strict';

    var self = this;

    console.info(
        'node redis rpc\n' +
        '\t>> on\n' +
        '\t\tchannel: %s\n' +
        '\t\thandler: %s\n' +
        '\t\tcallback: %s',
        channel, handler, callback
    );

    /**
     * Wrap the original event handler.
     *
     * Argument params:
     *
     * @param {Object} message
     * @param {String} channel
     * @param {Function} rpc this rpc callback handler will be added by this function
     */
    var wrappedHandler = function (/*message, channel*/) {
        var args = Array.prototype.slice.apply(arguments);
        var message = args[0];

        var rpcCallbackHandler = function () {};

        console.info(
            'node redis rpc\n' +
            '\t>> called extended wrapper\n' +
            '\t\ttype: "%s"\n' +
            '\t\tback channel: "%s"',
            message.__type,
            message.__backChannel
        );

        // Check if event is of type "rpc".
        if ('__type' in message && message.__type === 'rpc') {
            // After the application executes the "done" handler,
            // the result will be trigger back to the "back channel" (rpc response).
            // After that step, the "rpc callback" handler will handle the result data.
            rpcCallbackHandler = function (err, result) {
                self.emit(
                    message.__backChannel,
                    {
                        err: err,
                        result: result
                    }
                );
            };
        }

        // Append our extended rpc callback handler
        args = [].concat(args, [rpcCallbackHandler]);

        // Call original handler with extended args
        handler.apply(null, args);
    };

    // Trigger original method
    this._on(channel, wrappedHandler, callback);
};

module.exports = NodeRedisRpc;


benchmark/bs1.js
var Benchmark = require('benchmark');
var suite = new Benchmark.Suite;

//we have to globally declare the modules
global.NodeRedisRpc = require('../index');
global.NRP = require('node-redis-pubsub');
global.config = {
    host: 'localhost', // redis server hostname
    port: 6379,        // redis server port
    scope: 'test'      // use scope to prevent sharing messages between "node redis rpc"
};
global.nrp = new NRP(global.config);
global.nodeRedisRpcInst = new NodeRedisRpc(global.config);


global.myRpcCallback = function (err, result) {
    console.log('err', err);
    console.log('result', result);
};


//output helper functions
var startBench = function (event) {
    //console.log('bench "%s" started', event.target.name);
};
var cycleBench = function (event) {
    console.log('bench "%s" cycle %s', event.target.name, event.target.cycles);
};
var abortBench = function (event) {
    console.log('bench "%s" aborted', event.target.name);
};
var errorBench = function (event) {
    console.log('bench "%s" failed! Message: %s', event.target.name, event.message);
};
var resetBench = function (event) {
    console.log('bench "%s" resetted', event.target.name);
};
var completeBench = function (event) {
    //console.log('bench "%s" finished', event.target.name);
};


// define the benchmark suite with several benchmarks
suite
    // first benchmark the pub/sub package
    .add({
        'name': 'simple pub sub',
        'defer': true,
        'fn': function (deferred) {
            global.nrp.on('say_hello', function (data) {
                //console.log(data.name);
                //global.nrp.off('say_hello');
                deferred.resolve();
            });
            global.nrp.emit('say_hello', {name: 'Ray'});
        },
        'setup': function () {
        },
        'teardown': function () {
            global.nrp.off('say_hello');
            //global.nrp.quit();
        },
        'onStart': startBench,
        'onCycle': cycleBench,
        'onAbort': abortBench,
        'onError': errorBench,
        'onReset': resetBench,
        'onComplete': completeBench
    })

    // then benchmark the pub/sub behaviour with the rpc package on top
    .add({
        'name': 'simple pub sub with rpc package',
        'defer': true,
        'fn': function (deferred) {
            global.nodeRedisRpcInst.on('channel1', function (data, channel, done) {
                //console.log(data);
                deferred.resolve();
            });
            global.nodeRedisRpcInst.emit(
                'channel1',
                {channel: 'channel1', test: 'simple pub sub with rpc package'}
            );
        },
        'setup': function () {
        },
        'teardown': function () {
            global.nodeRedisRpcInst.off('channel1');
            //global.nodeRedisRpcInst.quit();
        },
        'onStart': startBench,
        'onCycle': cycleBench,
        'onAbort': abortBench,
        'onError': errorBench,
        'onReset': resetBench,
        'onComplete': completeBench
    })

    // then benchmark a simple rpc call
    .add({
        'name': 'simple rpc call',
        'defer': true,
        'fn': function (deferred) {
            global.nodeRedisRpcInst.on('channel1', function (data, channel, done) {
                //console.log(data);
                done(null, {foo: 'bar', num: 123});
            });
            global.nodeRedisRpcInst.emit(
                'channel1',
                {channel: 'channel1', test: 'simple rpc call'},
                {type: 'rpc', callback: function(){
                    deferred.resolve();
                }}
            );
        },
        'setup': function () {
        },
        'teardown': function () {
            global.nodeRedisRpcInst.off('channel1');
            //global.nodeRedisRpcInst.quit();
        },
        'onStart': startBench,
        'onCycle': cycleBench,
        'onAbort': abortBench,
        'onError': errorBench,
        'onReset': resetBench,
        'onComplete': completeBench
    })

    // add listeners
    .on('cycle', function (event) {
        console.log(String(event.target));
    })
    .on('complete', function () {
        console.log('Fastest is %s', this.filter('fastest').pluck('name'));
        console.log('Slowest is %s', this.filter('slowest').pluck('name'));
        console.log('Successful: %s', this.filter('successful').pluck('name'));
    })
    // run async
    .run({'async': true});